### ISSUE TITLE
[suggestion] Add a minimum word count for update to trigger

### ISSUE THREAD

**Issue by AndyScull** [JimmXinu/FanFicFare#130]
Hi. I think it would be useful both in manual update (with many links in batch) and in email-triggered update, so this setting should be either global, of present in both methods.
Basically, it should be there to prevent full-on 'Author Note' chapters - those are usually less than 1000 words, and contains only author words, either advertising his other stories, or saying the story is on hiatus... or just plain random stuff like a damn blog. Preferably length should be changeable, to fit everyone's needs.

Now, where this request comes from:
When manually updating, usually I check email and see how many words are there in new chapter, and if it's too short then I don't update story in calibre immediately. Now, I got tired of copy/pasting URLs from email and wanted to set up a auto-update from separate email account... but these short chapters would mess up everything if they're downloaded unconditionally. They will be downloaded, integrated into epub, and later any kind of update (except full relownload) would just skip this chapter, and author may later decide to actually add story content to it which I would totally miss.

**Comment by JimmXinu** [comment]
**TL;DR**: 

While not a bad idea, this isn't something that can be implemented very easily.  I'll leave this issue open for a while in case others want to chime in, but my initial reaction is that the added convenience in a few corner cases isn't worth the extra complexity and code it would require.

**Long answer:**

FanFicfare can find the number of 'chapters' in existing epubs by counting the chapter files.  There isn't any standard way to retrieve the existing word count.  While it may be present in title page, log page, or calibre custom columns, those are all configurable options that can vary from user to user.  I've investigated and even written code before to store all of FFF's metadata inside the epub, but there are too many ways for it to be discarded during other processing.  Plus, there are sites that don't report word counts at all.

So preventing update based on the number of new words is problematic.

Another possible option is to look at the length of 'new' content _after_ downloading and abort the update if it's under a certain limit.  

The network bandwidth and server time would still be expended, but it's somewhat more doable.  Word count would only be a rough estimate due to stripping HTML issues (some people get entirely too worked up about exact word counts).  Because it would be done after downloading, the calibre PI version would not be able to stop and ask the user to accept or reject the update.  And if you started reading a story of short chapters (100 word 'drabble' stories are a thing I've seen), you'd need to either change your settings or set specific lower limits for those stories.

(You might also ask your authors to stop posting non-chapters--as I recall, it's forbidden by the site terms on the most popular sites.  It's also pretty easy to delete a chapter from an epub using Calibre's Edit Book.)

**Comment by AndyScull** [comment]
Well, ok, you have some very good points in your reply. I'll think about possible ways to do it in my free time, can't think of anything right now.
One possible thing - maybe count not a number of words, but rather number of chars (after converting bytes->unicode) on downloaded chapter page? It would be very inaccurate and technical setting though. Maybe I'll just try to add couple of lines to plugin code on my computer and see how it goes...

And thanks for the tip, I'll try living in autodl mode with subsequent deletion of 'bad' chapters. 
Tried editing one epub, it seems deleting last xhtml file suffices, at least for my reader. Other readers may throw an error if there's still a reference in content.opf <metadata> section. I'm glad that FFF does not use some kind of metadata in xml to keep track of chapters and just counts a number of files

Do you think this suggestion should be closed now until some new ideas or should we let just it hang here for all to see?

**Comment by JimmXinu** [comment]
I don't mind leaving it open for a while.  I doubt it will garner a large audience, but surely even fewer people will look at a closed issue.

**Comment by MrTyton** [comment]
Solution might be to have configurable option to update anyways if chapters are the same but the modified dates are different.

From working the CLI, I've come across this:

`FFF: WARNING: 2016-09-15 13:42:11,777: base_writer.py(195): File(Spectrum-ffnet_12102203.epub) Updated(2016-09-15) more recently than Story(2016-09-12) - Skipping`

so there is knowledge about when the epub has been updated compared to when the story has been. If the story has been updated on site (on say 2016-09-16) and my file has been last updated earlier (2016-09-15) then we can do an overwrite of either the last chapter or the entire epub if the number of chapters are the same.

Does it make sense to do that?

**Comment by JimmXinu** [comment]
The plugin version has an option like that: 'Overwrite if Newer', and in CLI, there's an INI option `always_overwrite` which defaults to false (and gives the warning you saw).
```
## Always overwrite output files.  Otherwise, the downloader checks
## the timestamp on the existing file and only overwrites if the story
## has been updated more recently.  Command line version only
#always_overwrite: true
```
The downside is most sites only give an updated _date_, not time.  So FFF only stops the overwrite if the file date is _greater than_ the updated date.  If you run it more than once on the day the update comes out, it will run the full download and overwrite more than once.

Epub updates (-u/--update-epub) causes chapter count to take precedence over the file date vs updated date.

If updated date is more recent than file date while chapter count is unchanged may indicate changes, but we've no way of knowing where (which chapters).  You can manually run download again without update in that case.

None of which helps with the original request, which was about skipping super-short chapters on the theory they were author blather instead of content.

At this point, I'm going to declare the discussion closed.  If anything, I'm less inclined to try to implement it than I was before.  Sorry, AndyScull.

**Comment by MrTyton** [comment]
Good point, I forgot about that option.
